### PR TITLE
WIP make it a CLJC library

### LINKED ISSUE #8: Make it possible to use DartClojure from ClojureScript
I want to add a Dart->Clojure converter to Calva:

* https://github.com/BetterThanTomorrow/calva/issues/1763

The way easiest way to do this would be that DartClojure was made to work with ClojureScript, because Calva code can then do the conversion.

I have forked and am converting things to `cljc`. Of course running into many interesting problems, but I am making some progress. 😄 Are you interested in a PR?

### PR DESCRIPTION
**Pull request by PEZ**
I've gotten a POC to work with Calva. See https://twitter.com/pappapez/status/1534654693338759169 for a demo.

Seems I have messed something up with the `$` string substitution. Because that test now fails. I'll see if I can spot what it could be about. I did have some troubles making the `clean` function to work in ClojureScript and in making it work I guess the Clojure behaviour fell victim...

Also, thinking that maybe we should add shadow-cljs build help for the development. Right now it is a bit of loading all-the-files a lot when working with the cljs REPL.

Also, probably should make the tests CLJC as well.

Let me know what you think!

Fixes #8

### CODE REVIEW

**Comment by PEZ** [comment]
I started with adding cljs-test-runner to the mix. (I do think that if we go for shadow-cljs for the dev workflow, maybe use it for the tests as well, but wanted to give this a shot first.)


The tests results for ClojureScript are a bit less great than for Clojure right now, but maybe it is just one or two places where I need to see what data type I have, since I don't seem to have a string:

```
~/Projects/DartClojure(pez/cljc↑1|✔) % clj -M:test-cljs

Testing dumch.improve-test

Testing dumch.integration-test

ERROR in (complex-example-test2) (/Users/pez/Projects/DartClojure/test/dumch/integration_test.cljc:82:7)
dart->clojure, using nest macro
expected: (= (-> code2 convert read-string) (quote (m/Column :children [(m/Question (get (get questions _questionIndex) "questionText")) (-> (get (get questions _questionIndex) "answers") (.map (fn [answer] (m/Answer _answerQuestion answer))) (.toList)) (m/Numb :sorted (and (= a b) (< 1 2 3 a)))])))
  actual: #object[TypeError TypeError: s.replace is not a function]

Testing dumch.parse-test

ERROR in (logic-test) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:173:9)
and: &&
expected: (= (dart->clojure "b && a") (quote (and b a)))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (logic-test) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:175:11)
or: ||, ??
expected: (= (dart->clojure "b || a") (quote (or b a)))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (math-test) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:237:9)
sum
expected: (= (dart->clojure "b + a") (quote (+ b a)))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (math-test) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:245:9)
multiply
expected: (= (dart->clojure "b * a") (quote (* b a)))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (lambda-test) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:249:9)
lambda with =>
expected: (= (dart->clojure "Button(onPressed: (ctx) => 1 + 1;)") (quote (Button :onPressed (fn [ctx] (+ 1 1)))))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (lambda-test) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:255:9)
lambda with typed argument
expected: (= (dart->clojure "Button(onPressed: (Context ctx) => 1 + 1;)") (quote (Button :onPressed (fn [ctx] (+ 1 1)))))
  actual: #object[TypeError TypeError: s.replace is not a function]

FAIL in (strings-test) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:315:9)
$ substitution
expected: (= (dart->clojure "Text('Some $field and ${Factory.create()}')") (quote (Text (str "Some " field " and " (.create Factory)))))
  actual: (not (= (Text (str "Some " field "field and " (.create Factory) "Factory.create()")) (Text (str "Some " field " and " (.create Factory)))))

ERROR in (flatten-same-operators) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:349:9)
calculation results
expected: (-> "(2 + 1 + 1 == 4 && true) && (1 == 1 || false) && 3 == 3" dart->clj-string read-string eval)
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (flatten-same-operators) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:355:11)
+, *, and, or
expected: (= (-> (str 1 sym 2 sym 3 sym 4 sym 5) dart->clojure) (clojure.core/sequence (clojure.core/seq (clojure.core/concat (clojure.core/list (symbol (case sym "||" (quote or) "&&" (quote and) sym))) (clojure.core/list 1) (clojure.core/list 2) (clojure.core/list 3) (clojure.core/list 4) (clojure.core/list 5)))))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (flatten-same-operators) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:355:11)
+, *, and, or
expected: (= (-> (str 1 sym 2 sym 3 sym 4 sym 5) dart->clojure) (clojure.core/sequence (clojure.core/seq (clojure.core/concat (clojure.core/list (symbol (case sym "||" (quote or) "&&" (quote and) sym))) (clojure.core/list 1) (clojure.core/list 2) (clojure.core/list 3) (clojure.core/list 4) (clojure.core/list 5)))))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (flatten-same-operators) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:355:11)
+, *, and, or
expected: (= (-> (str 1 sym 2 sym 3 sym 4 sym 5) dart->clojure) (clojure.core/sequence (clojure.core/seq (clojure.core/concat (clojure.core/list (symbol (case sym "||" (quote or) "&&" (quote and) sym))) (clojure.core/list 1) (clojure.core/list 2) (clojure.core/list 3) (clojure.core/list 4) (clojure.core/list 5)))))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (flatten-same-operators) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:355:11)
+, *, and, or
expected: (= (-> (str 1 sym 2 sym 3 sym 4 sym 5) dart->clojure) (clojure.core/sequence (clojure.core/seq (clojure.core/concat (clojure.core/list (symbol (case sym "||" (quote or) "&&" (quote and) sym))) (clojure.core/list 1) (clojure.core/list 2) (clojure.core/list 3) (clojure.core/list 4) (clojure.core/list 5)))))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (flatten-same-operators) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:360:9)
flatten compare operators when possible
expected: (= (-> "3 > 2 && 2 > 1 && 4 > 3 && true" dart->clj-string) "(and true (> 4 3 2 1))")
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (flatten-same-operators) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:362:9)
flatten compare operators when possible
expected: (= (-> "4 >= 3 && 3 >= 2 || 2 >= 1" dart->clojure) (quote (and (>= 4 3) (or (>= 3 2) (>= 2 1)))))
  actual: #object[TypeError TypeError: s.replace is not a function]

ERROR in (flatten-same-operators) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:364:9)
flatten compare operators when possible
expected: (= (-> "5 >= 4 && 4 >= 3 && 3 >= 2 || 2 >= 1" dart->clojure) (quote (and (or (>= 3 2) (>= 2 1)) (>= 5 4 3))))
  actual: #object[TypeError TypeError: s.replace is not a function]

Ran 29 tests containing 106 assertions.
1 failures, 15 errors.
```

File: src/dumch/parse.cljc [hunk 1]
```diff
@@ -287,38 +305,47 @@
         (str/replace #"(\s*@.*\n)" "\n")
 
         ;; cleaning code from comments
-        (str/replace #"\/\*(\*(?!\/)|[^*])*\*\/" "")    ; /* ... */ 
-        (str/replace #"(\/\/).*" "")                    ; // ... 
-        (str/replace str-pattern 
-                     (transform (comp substitute-curly-quotes 
-                                      decode)))))) 
+        (str/replace #"/\*(\*(?!/)|[^*])*\*/" "")    ; /* ... */ 
+        (str/replace #"(//).*" "")                   ; // ...
```
Comment (Liverm0r) [inline src/dumch/parse.cljc:309]: I think javascript wont work without escaping 
image
https://regex101.com/

File: src/dumch/util.cljc [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+(def nl (n/newlines 1))
+(defn mapn [f sq] (interpose nl (map f sq)))
+
+(defn with-print [arg]
+  (println :with-print arg)
+  arg)
+
+#?(:clj
+   (defmacro inline-resource [resource-path]
+     (slurp (clojure.java.io/resource resource-path))))
```
[the remark this answers is not in the document]
Comment (PEZ) [reply src/dumch/util.cljc:21]: It's basically this recipe: https://code.thheller.com/blog/shadow-cljs/2019/10/12/clojurescript-macros.html

Though we are using a cljc file instead, which can be regarded as one cljs file and one clj file, where the cljs one requires the macros of the clj one. The end result is that the resource will be inlined in the ClojureScript version of the compiled library.

Does this make sense?

File: src/dumch/parse.cljc (hunk 1 above)
Comment (PEZ) [reply src/dumch/parse.cljc:309]: It wouldn't compile in ClojureScript with the escaping, actually. See https://clojurians.slack.com/archives/C03S1L9DN/p1654705094721129

I think it does the right thing in both Clojure and ClojureScript now, but then again, we do have failing tests... I haven't figured those out yet.

File: src/dumch/parse.cljc [hunk 3]
```diff
@@ -38,41 +44,52 @@
...
-
-(defn- substitute-$ 
+           (cond open-b? (doto stack (.push 1))
+                 close-b? (doto stack .pop)
+                 :else stack)
+           (if expr-ends? (conj rslt #?(:clj (.toString cur-expr)
+                                        :cljs (.join cur-expr ""))) rslt)
+           cur-str
+           #?(:clj (if expr-ends? (StringBuilder.) (.append cur-str ch))
+              :cljs (if expr-ends? (js/Array.) (doto cur-str (.push ch))))
```
[the remark this answers is not in the document]
Comment (PEZ) [reply src/dumch/parse.cljc]: Ah, thanks! This fixed the failing test for both Clojure and ClojureScript.

[the remark this answers is not in the document]
Comment (PEZ) [reply src/dumch/parse.cljc]: There are erroring tests for the ClojureScript part. 15 when I wrote that comment. With the latest commit we are down to 1.

**Comment by PEZ** [comment]
Found the source of 14 of those ClojureScript test errors. The remaining one could prove a challenge I guess:

```
~/Projects/DartClojure(pez/cljc|✚1) % clj -X:test     

Running tests in #{"test"}

Testing dumch.improve-test

Testing dumch.integration-test

Testing dumch.parse-test

Ran 29 tests containing 106 assertions.
0 failures, 0 errors.

~/Projects/DartClojure(pez/cljc|✔) % clj -M:test-cljs

Testing dumch.improve-test

Testing dumch.integration-test

Testing dumch.parse-test

ERROR in (flatten-same-operators) (/Users/pez/Projects/DartClojure/test/dumch/parse_test.cljc:349:9)
calculation results
expected: (-> "(2 + 1 + 1 == 4 && true) && (1 == 1 || false) && 3 == 3" dart->clj-string read-string eval)
  actual: #object[Error Error: cljs.core/*eval* not bound]

Ran 29 tests containing 106 assertions.
0 failures, 1 errors.
```

File: src/dumch/parse.cljc [hunk 4]
```diff
@@ -247,7 +248,7 @@
     :return (if v1 (ast->clj v1) (tnode 'nil))
     :typecasting (ast->clj v1)
     :const (n/meta-node (tnode :const) (ast->clj v1))
-    :identifier (symbol (str/replace v1 #"!" ""))
+    :identifier (symbol (str/replace (str v1) #"!" ""))
```
Comment (Liverm0r) [inline src/dumch/parse.cljc:251]: I think the only source of symbols here is the `flatten-commutative-node` method.  Lets just change it to
```clojure
(defn- flatten-commutative-node [[f :as node]]
  (list* 
    [:identifier (case f :and "and", :or "or", :add "+", :mul "*")]
    (flatten-same-node node)))
```

File: src/dumch/parse.cljc [hunk 5]
```diff
@@ -134,68 +151,69 @@
...
             compare-ast (build-ast compare-vals)]
-        (if (= (count and-node) 1) 
+        (if (= (count and-node) 1)
           (next compare-ast)
           (conj and-node compare-ast)))
       and-node)))
 
 (defn ast->clj [[tag v1 v2 v3 v4 :as node]]
   #_(println :node node)
+  (def v1 v1)
```
Comment (Liverm0r) [inline src/dumch/parse.cljc:182]: unnecessary def

**Comment by Liverm0r** [comment]
lets ignore this last test, I think

**Review by Liverm0r — APPROVED** [review]

**Pull request merged**